### DISCUSSION TITLE
executing job failed: Argument list too long

### DISCUSSION THREAD

**Discussion by Shamaoke** [junegunn/fzf#3402] [Q&A] [answered]
Hi.

I'm writing a script¹ that should display a list of entries for selection in Fzf that is running in a Vim pop-up window. The entries themselves are read from a text file. The problem is that in a case of a large number of entries the program crashes, returning the error: `executing job failed: Argument list too long`.

What happens and why? How the mentioned problem could be solved?

Thanks.

¹[fzf-numbers.vim](https://gist.github.com/Shamaoke/4f0f7675fa5fd70dbcb5c405982dec04)

**Comment by junegunn** [comment]
It's an error from your shell, not fzf. There is a limit to the number/length of arguments you can pass to a command.

* https://stackoverflow.com/questions/11289551/argument-list-too-long-error-for-rm-cp-mv-commands

It looks like `echo` isn't needed at all. Why don't you run `cat data.txt | fzf`?

**Comment by Shamaoke** [comment]
> It looks like echo isn't needed at all. Why don't you run cat data.txt | fzf?

I gave a simplified version of the script. In the original version, after reading data from the file, it's further processed and formatted. Only after that the resulting data is passed to the input of Fzf. I made a [change](https://gist.github.com/Shamaoke/4f0f7675fa5fd70dbcb5c405982dec04#file-fzf-numbers-vim-L67) to the sample script to illustrate the matter.

It looks like, this resulting data should also be saved in a file. After that the file should be read in Fzf command, as you wrote, i. e. `$FZF_DEFAULT_COMMAND = $"cat result.txt"`.

If you have any other thoughts on the problem, I would be glad to hear them.

**Reply by junegunn** [reply, marked as answer]
This is a limitation of your shell environment, there's nothing fzf can help alleviate the problem. I mean, you'll run into the same problem just by running the `echo` command.

 You'll either have to save the result to a temp file or find another way to filter/transform the list. e.g. `seq 10000 | perl -ne 'print if $_ % 2 == 0' | fzf`

**Reply by junegunn** [reply]
FWIW, it is not recommended to temporarily change `$FZF_DEFAULT_COMMAND` as the variable is propagated to the child processes which may lead to unexpected results. https://github.com/junegunn/fzf/commit/7374fe73a3e91bafa5b27b5781464f16a63e385d

**Comment by Shamaoke** [comment]
> This is a limitation of your shell environment, there's nothing fzf can help alleviate the problem.

Understood. The issue is closed.

> FWIW, it is not recommended to temporarily change $FZF_DEFAULT_COMMAND as the variable is propagated to the child processes which may lead to unexpected results.

As far as I understand, this is the only way to pass a command to Fzf, that it should execute when it starts. In [skim](https://github.com/lotabout/skim), an alternative to Fzf, there is a special option for that matter, i. e. `sk --cmd='seq 1 1000'`. Fzf doesn't have this and so, you have to use environment variable tricks when writing Fzf based Vim plugins.

**Comment by junegunn** [comment]
> this is the only way to pass a command to Fzf

No, you can just run the source command with fzf using a shell.

```vim
call term_start(['sh', '-c', 'seq 100 | fzf'])
```

> `sk --cmd='seq 1 1000'`

Why do we need it when we can `seq 1 1000 | sk`? It's shorter and it's using the standard shell syntax everyone already knows.

FWIW, fzf can also do that using a far more extensible event-binding mechanism.

```sh
: | fzf --bind 'start:reload(seq 1 1000)'
```

More examples on https://github.com/junegunn/fzf/blob/master/ADVANCED.md

**Comment by Shamaoke** [comment]
>  `fzf --bind 'start:reload(seq 1 1000)'`

Great! I [rewrote](https://gist.github.com/Shamaoke/7e88ef6922b7c5d4e2444991c21475c4) the script in accordance with your recommendations. In addition, in the new version of the script, the processed data is written to a file and subsequently read from it. Therefore, no error occurs. I tested the script on 1,000,000 entries and the program works with acceptable performance for me. Is it something else, I need to care about?

> More examples on https://github.com/junegunn/fzf/blob/master/ADVANCED.md

That's an interesting read. Thanks!

**Reply by junegunn** [reply]
> Is it something else, I need to care about?

Not much. Oh, I would use `--border` option instead of setting the popup border options.

**Comment by Shamaoke** [comment]
> I would use --border option instead of setting the popup border options.

With the `--border` option you can also use the `--border-label` option. For what reason a field index expression can't be used in this option (e. g. `--border --border-label=Number {1} is selected`). Are these conceptual reasons or technical difficulties?

**Reply by junegunn** [reply]
`--border-label` wasn't designed to support expressions that need to be dynamically evaluated. If you want it, use `transform-border-label` action.

See https://github.com/junegunn/fzf/blob/master/CHANGELOG.md#0370

**Comment by Shamaoke** [comment]
````
fzf \
  --border=sharp \
  --bind='focus:transform-border-label:printf "Number %4d is selected" {1}' \
  --bind='start:reload:seq 1 1000'
````
Perfect!

Thank you very much for your explanations and for the help!
